x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/CopySign.java

```diff
@@ -0,0 +1,190 @@
+    }
+
+    private CopySign(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(children().get(0));
+        out.writeNamedWriteable(children().get(1));
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, CopySign::new, children().get(0), children().get(1));
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        if (newChildren.size() != 2) {
+            throw new EsqlIllegalArgumentException("Function [{}] expects exactly two arguments, got [{}]", NAME, newChildren.size());
+        }
+        return new CopySign(source(), newChildren.get(0), newChildren.get(1));
+    }
+
+    @Override
+    public DataType dataType() {
+        if (dataType == null) {
+            resolveType();
+        }
+        return dataType;
+    }
+
+    @Override
+    public TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+        var magnitude = children().get(0);
+        var sign = children().get(1);
+        if (magnitude.dataType().isNumeric() == false) {
+            return new TypeResolution("Magnitude must be a numeric type");
+        }
+        if (sign.dataType().isNumeric() == false) {
+            return new TypeResolution("Sign must be a numeric type");
+        }
+        // The return type is the same as the magnitude type, so we can use it directly.
+        dataType = magnitude.dataType();
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public boolean foldable() {
+        return Expressions.foldable(children());
+    }
+
+    @Override
+    public EvalOperator.ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
+        var dataType = dataType();
+        if (FACTORY_PROVIDERS.containsKey(dataType) == false) {
+            throw new EsqlIllegalArgumentException("Unsupported data type [{}] for function [{}]", dataType(), NAME);
+        }
+        var sign = children().get(1);
+        var magnitude = children().get(0);
+        return FACTORY_PROVIDERS.get(dataType)
+            .create(
+                source(),
+                toEvaluator.apply(magnitude),
+                // We always cast the sign to double for processing, as the sign can be any numeric type.
+                Cast.cast(source(), sign.dataType(), DataType.DOUBLE, toEvaluator.apply(sign))
+            );
+    }
+
+    @Evaluator(extraName = "Float")
+    static float processFloat(float magnitude, double sign) {
```

Comment:
Should this one be `float, float`? For most of our numeric things we tend to use the casing logic to promote small things to big. I'd *expect* the `float, float` and `double, double` and `long, long` and `int, int` and then use `Cast.cast` to promote the "smaller" side. We certainly *could* do the full combinatorial explosion, but we haven't before and I wouldn't start on COPY_SIGN.